sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/StorageHelper.cs

```diff
@@ -0,0 +1,114 @@
+                    {
+                        dirPath = CreateTelemetryDirectory(localAppData);
+                        if (dirPath != null)
+                        {
+                            defaultStorageDirectory = dirPath;
+                            return defaultStorageDirectory;
+                        }
+
+                        string temp = environmentVars["TEMP"].ToString();
+                        if (temp != null)
+                        {
+                            dirPath = CreateTelemetryDirectory(temp);
+                            if (dirPath != null)
+                            {
+                                defaultStorageDirectory = dirPath;
+                                return defaultStorageDirectory;
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    string tmpdir = environmentVars["TMPDIR"].ToString();
+                    if (tmpdir != null)
+                    {
+                        dirPath = CreateTelemetryDirectory(tmpdir);
+                        if (dirPath != null)
+                        {
+                            defaultStorageDirectory = dirPath;
+                            return defaultStorageDirectory;
+                        }
+                    }
+
+                    dirPath = CreateTelemetryDirectory(nonWindowsVarTmp);
+                    if (dirPath != null)
+                    {
+                        defaultStorageDirectory = dirPath;
+                        return defaultStorageDirectory;
+                    }
+
+                    dirPath = CreateTelemetryDirectory(nonWindowsTmp);
+                    if (dirPath != null)
+                    {
+                        defaultStorageDirectory = dirPath;
+                        return defaultStorageDirectory;
+                    }
+                }
+
+                return defaultStorageDirectory;
+            }
+        }
+
+        /// <summary>
+        /// Creates directory for storing telemetry.
+        /// </summary>
+        /// <param name="path">Base directory.</param>
+        /// <returns>Directory path if it is created else null.</returns>
+        private static string CreateTelemetryDirectory(string path)
+        {
+            try
+            {
+                string telemetryDirPath = Path.Combine(path, "Microsoft", "ApplicationInsights");
+                Directory.CreateDirectory(telemetryDirPath);
+                return telemetryDirPath;
+            }
+            catch (Exception ex)
+            {
+                AzureMonitorExporterEventSource.Log.Write($"ErrorCreatingDefaultStorageFolder{EventLevelSuffix.Error}", $"{ex.ToInvariantString()}");
+                return null;
+            }
+        }
+
+        internal static bool IsWindowsOS()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
Could this be cached somewhere?

Reply:
This will get called once only (Except when we are not able to create subdir at any of the tried locations). We cache defaultStorageDirectory path once it is initialized. (line 19).